urlc175f05cb8_mchllbrt_PermCode_tgz-pJ8-permlib_examples_UniversalSeparableJ8/src/permlib/utilities/RestrictedPermutation.java

```diff
@@ -0,0 +1,509 @@
+     * @return modified restricted permutation
+     */
+    public RestrictedPermutation deletePointAt(int index) {
+        RestrictedPermutation result = this.copy();
+        result.p = PermUtilities.delete(p, index);
+        result.permString = result.p.toString();
+        result.createForbiddenPairs();
+        result.userForbiddenPairs.clear();
+        for (int i = 0; i <= result.p.length(); i++) {
+            for (int j = 0; j <= result.p.length(); j++) {
+                boolean isForbidden = true;
+                isForbidden &= userForbiddenPairs.contains(new IntPair(i + (i > index ? 1 : 0), j + (j > p.elements[index] ? 1 : 0)));
+                if (i == index) {
+                    isForbidden &= userForbiddenPairs.contains(new IntPair(i + 1, j + (j > p.elements[index] ? 1 : 0)));
+                }
+                if (j == index) {
+                    isForbidden &= userForbiddenPairs.contains(new IntPair(i + (i > index ? 1 : 0), j + 1));
+                }
+                if (i == index && j == index) {
+                    isForbidden &= userForbiddenPairs.contains(new IntPair(i + 1, j + 1));
+                }
+                if (isForbidden) {
+                    result.userForbiddenPairs.add(new IntPair(i, j));
+                }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Returns a copy of the restricted permutation. Note, this performs a deep
+     * copy.
+     *
+     * @return a copy of the restricted permutation
+     */
+    private RestrictedPermutation copy() {
+        RestrictedPermutation result = new RestrictedPermutation(basisString, permString, simple, involution);
+        result.p = this.p.clone();
+        result.forbiddenPairs = new HashSet<IntPair>(this.forbiddenPairs);
+        result.properties = new ArrayList<PermProperty>(this.properties);
+        result.userForbiddenPairs = new HashSet<IntPair>(this.userForbiddenPairs);
+        return result;
+    }
+
+    /**
+     * Adds a property to the basis.
+     */
+    private void addBasisProperties() {
+        Scanner basisElements = new Scanner(basisString);
+        while (basisElements.hasNextLine()) {
+            String line = basisElements.nextLine();
+            if (line.length() > 0) {
+                properties.add(PermUtilities.avoidanceTest(line));
+            }
+        }
+    }
+
+    /**
+     * Checks whether newly created gaps should be marked as forbidden based on
+     * user restrictions. That is, if both neighbouring positions along the
+     * split are marked forbidden then so will the new space, otherwise not.
+     *
+     * @param r old row index
+     * @param c old column index
+     * @param i new row index
+     * @param j new column index
+     * @param length the length of the inserted permutation
+     * @return true if new space should be forbidden, false otherwise
+     */
+    private boolean shouldForbid(int r, int c, int i, int j, int length) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 488-502 into a separate method.